### DISCUSSION TITLE
Preview and display of pinocchio::Model

### DISCUSSION THREAD

**Discussion by kozakromch** [stack-of-tasks/pinocchio#1497] [Ideas]
Hello, I'm using you library without any visual geometry, and it is little hard to validate model. 
So what do you think about realization some few functions: 

1) function that create a geometry model based on joints and rigid bodies. It will create a sticks from parent joint to its child and ellipsoid from rigid bodies. 
2) function that use smth like boost graph vizualizer. It will create a image with a topology of model ( Need GraphViz to be added) add some additional info could be added in nodes
3) some additional function that just print out some info about model. (I know it simple, but with this lib will be a little more user friendly)

I guess all this function I could realize by myself, but sometimes I'll need you help. 
So what do you think? Does this function need to be implemented in this library ?

**Comment by jcarpent** [comment]
>  1. function that create a geometry model based on joints and rigid bodies. It will create a sticks from parent joint to its child and ellipsoid from rigid bodies.



Do you know that Pinocchio is coming with visualisation features in Python mostly? Please refer to these lines:
https://github.com/stack-of-tasks/pinocchio#visualization

**Reply by kozakromch** [reply]
For my project I planned to create a model by c++ ( as it's a project language) than save it to XML and load load by python and after that in python create geometries and visualize it. But, I guess, it's not the good way for all users. 
Do you have more convenient variants for this?

**Reply by jcarpent** [reply]
You can create a model in Python and load it in C++ via the Python parser.

**Comment by jcarpent** [comment]
>  2\. function that use smth like boost graph vizualizer. It will create a image with a topology of model ( Need GraphViz to be added) add some additional info could be added in nodes

This feature can be indeed useful to display the tree structure of a system.
This is more than welcome.

**Comment by jcarpent** [comment]
>   3\. some additional function that just print out some info about model. (I know it simple, but with this lib will be a little more user friendly)

Could you elaborate a bit more with examples?

**Reply by kozakromch** [reply]
Well, It could be a simple method in ModelTPL structure that give's something like that result: \
`
Model name :  model_name \
Number of joints: 30 \
Number of bodies: 25 \
Number of frame: 30 \
NQ: 60 \
NV: 55 \
Names of joints: "pelvis", "foot_l", "foot_r" ...\
Parent child connectivity: "Pelvis" -> "Chest", "Pelvis" ->"Hip_l", "Pelvis" ->"Hip_r" ... \
And so on \
`

**Reply by wxmerkt** [reply]
@kozakromch Something like this already exists, e.g.:

```python
import pinocchio as pin
model = pin.buildSampleModelHumanoid()
print(model)
```
outputs:
```
Nb joints = 30 (nq=35,nv=34)
  Joint 0 universe: parent=0
  Joint 1 freeflyer_joint: parent=0
  Joint 2 rleg_shoulder1_joint: parent=1
  Joint 3 rleg_shoulder2_joint: parent=2
  Joint 4 rleg_shoulder3_joint: parent=3
  Joint 5 rleg_elbow_joint: parent=4
  Joint 6 rleg_wrist1_joint: parent=5
  Joint 7 rleg_wrist2_joint: parent=6
  Joint 8 lleg_shoulder1_joint: parent=1
  Joint 9 lleg_shoulder2_joint: parent=8
  Joint 10 lleg_shoulder3_joint: parent=9
  Joint 11 lleg_elbow_joint: parent=10
  Joint 12 lleg_wrist1_joint: parent=11
  Joint 13 lleg_wrist2_joint: parent=12
  Joint 14 chest1_joint: parent=1
  Joint 15 chest2_joint: parent=14
  Joint 16 head1_joint: parent=15
  Joint 17 head2_joint: parent=16
  Joint 18 rarm_shoulder1_joint: parent=15
  Joint 19 rarm_shoulder2_joint: parent=18
  Joint 20 rarm_shoulder3_joint: parent=19
  Joint 21 rarm_elbow_joint: parent=20
  Joint 22 rarm_wrist1_joint: parent=21
  Joint 23 rarm_wrist2_joint: parent=22
  Joint 24 larm_shoulder1_joint: parent=15
  Joint 25 larm_shoulder2_joint: parent=24
  Joint 26 larm_shoulder3_joint: parent=25
  Joint 27 larm_elbow_joint: parent=26
  Joint 28 larm_wrist1_joint: parent=27
  Joint 29 larm_wrist2_joint: parent=28
```

**Reply by kozakromch** [reply]
Wow, thanks, I didn't notice that

**Comment by jcarpent** [comment]
Thanks @kozakromch for starting this discussion.
I hope you and other users/contributors would be able to animate this discussion.
